### PR TITLE
feat: performance improvement

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

This pr is to improve the performance of greenfield.

### Changes

Notable changes:
* add prefetch to speed up block
* add signature cache
* add block transaction numbers limit
* modify db

### CODE REVIEW

File: baseapp/abci.go [hunk 1]
```diff
@@ -504,7 +504,127 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+
+	gasMeter := app.getBlockGasMeter(app.preDeliverState.ctx)
+
+	app.preDeliverState.ctx = app.preDeliverState.ctx.
+		WithBlockGasMeter(gasMeter).
+		WithHeaderHash(req.Hash).
+		WithConsensusParams(app.GetConsensusParams(app.preDeliverState.ctx))
+
+	// call the hooks with the BeginBlock messages
+	for _, streamingListener := range app.abciListeners {
```
Comment (guagualvcha) [inline baseapp/abci.go:544]: Too many useless code.

File: baseapp/abci.go [hunk 2]
```diff
@@ -504,7 +504,127 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+			panic(fmt.Errorf("BeginPreBlock listening hook failed, height: %d, err: %w", req.Header.Height, err))
+		}
+	}
+
+	return res
+}
+
+func (app *BaseApp) PreDeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliverTx) {
+	defer func() {
+		for _, streamingListener := range app.abciListeners {
```
Comment (guagualvcha) [inline baseapp/abci.go:555]: useless code

File: baseapp/abci.go [hunk 3]
```diff
@@ -504,7 +504,127 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+	// The write to the DeliverTx state writes all state transitions to the root
+	// MultiStore (app.cms) so when Commit() is called is persists those values.
+	app.preDeliverState.ms.Write()
+
+	res := abci.ResponseCommit{
+		RetainHeight: retainHeight,
+	}
+
+	// call the hooks with the Commit message
+	for _, streamingListener := range app.abciListeners {
```
Comment (guagualvcha) [inline baseapp/abci.go:589]: Too many useless code.

File: baseapp/abci.go [hunk 4]
```diff
@@ -504,7 +504,68 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
 }
 
-// halt attempts to gracefully shutdown the node via SIGINT and SIGTERM falling
+// PreBeginBlock implements the ABCI application interface.
+func (app *BaseApp) PreBeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeginBlock) {
+	if req.Header.ChainID != app.chainID {
+		panic(fmt.Sprintf("invalid chain-id on BeginBlock; expected: %s, got: %s", app.chainID, req.Header.ChainID))
+	}
+
+	if err := app.validateHeight(req); err != nil {
```
Comment (guagualvcha) [inline baseapp/abci.go:513]: Please ensure that all these `pre` function will never panic, or even it panic, we should recover it and never let it crash the whole  chain node.

Comment (guagualvcha) [reply baseapp/abci.go:513]: The check here is unnecessary.

File: baseapp/abci.go (hunk 1 above)
Comment (Pythonberg1997) [reply baseapp/abci.go]: Removed

File: baseapp/abci.go (hunk 2 above)
Comment (Pythonberg1997) [reply baseapp/abci.go]: Removed

File: baseapp/abci.go (hunk 3 above)
Comment (Pythonberg1997) [reply baseapp/abci.go]: Removed

File: baseapp/abci.go (hunk 4 above)
Comment (Pythonberg1997) [reply baseapp/abci.go]: Removed

File: baseapp/abci.go [hunk 5]
```diff
@@ -504,7 +503,79 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+}
+
+func (app *BaseApp) PreDeliverTx(req abci.RequestPreDeliverTx) (res abci.ResponsePrefetch) {
+	defer func() {
+		if r := recover(); r != nil {
+			res = abci.ResponsePrefetch{Error: errorsmod.Wrapf(sdkerrors.ErrPanic, "%v", r).Error()}
+		}
+	}()
+
+	preState := app.preDeliverStates[req.StateIndex]
```
Comment (guagualvcha) [inline baseapp/abci.go:543]: race condition here.

File: baseapp/abci.go [hunk 6]
```diff
@@ -504,7 +503,79 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+		err := fmt.Sprintf("invalid request chain-id: %s, expected: %s", req.Header.ChainID, app.chainID)
+		return abci.ResponsePrefetch{Error: err}
+	}
+
+	if app.cms.TracingEnabled() {
+		app.cms.SetTracingContext(map[string]interface{}{"blockHeight": req.Header.Height})
+	}
+
+	// Initialize the preDeliverTx state.
+	app.setPreState(req.StateIndex, req.Header)
```
Comment (guagualvcha) [inline baseapp/abci.go:524]: race condition here.

File: baseapp/abci.go [hunk 7]
```diff
@@ -504,7 +503,79 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+
+func (app *BaseApp) PreCommit(req abci.RequestPreCommit) (res abci.ResponsePrefetch) {
+	defer func() {
+		if r := recover(); r != nil {
+			res = abci.ResponsePrefetch{Error: errorsmod.Wrapf(sdkerrors.ErrPanic, "%v", r).Error()}
+		}
+	}()
+
+	app.preDeliverStates[req.StateIndex].ms.Write()
+	app.preDeliverStates[req.StateIndex] = nil // empty/reset the pre-deliver state
```
Comment (guagualvcha) [inline baseapp/abci.go:572]: Let us do this on beginning blocks.

**Review by guagualvcha — APPROVED** [review]

File: x/auth/signing/verify.go [hunk 8]
```diff
@@ -53,6 +63,11 @@ func VerifySignature(pubKey cryptotypes.PubKey, signerData SignerData, sigData s
 			if !pubKey.Equals(pk) {
 				return errorsmod.Wrapf(sdkerrors.ErrorInvalidSigner, "feePayer's pubkey %s is different from signature's pubkey %s", pubKey, pk)
 			}
+
+			// add the tx to the cache if needed
+			if sigCache != nil && txBytes != nil {
+				sigCache.Add(string(txBytes), tx)
```
Comment (forcodedancing) [inline x/auth/signing/verify.go:69]: []byte could be big. It is ok to use `sigCache.Add( keeck256hash (txBytes)` ?

File: baseapp/abci.go [hunk 9]
```diff
@@ -504,7 +503,65 @@ func (app *BaseApp) Commit() abci.ResponseCommit {
...
+}
+
+func (app *BaseApp) PreCommit(req abci.RequestPreCommit) (res abci.ResponsePrefetch) {
+	defer func() {
+		if r := recover(); r != nil {
+			res = abci.ResponsePrefetch{Error: errorsmod.Wrapf(sdkerrors.ErrPanic, "%v", r).Error()}
+		}
+	}()
+
+	app.preDeliverStates[req.StateIndex].ms.Write()
```
Comment (forcodedancing) [inline baseapp/abci.go:558]: Why do need to write? How precommit (writing) will improve performance?

**Pull request merged**